crates/bevy_render/src/allocator.rs

```diff
@@ -0,0 +1,563 @@
+//! allocations get their own individual slabs. Due to platform limitations, on
+//! WebGL 2 all vertex buffers are considered large allocations that get their
+//! own slabs.
+//!
+//! The purpose of packing meshes together is to reduce the number of times
+//! vertex and index buffers have to be re-bound, which is expensive.
+//!
+//! [`offset-allocator`]: https://github.com/pcwalton/offset-allocator/
+//! [Sebastian Aaltonen's `OffsetAllocator`]: https://github.com/sebbbi/OffsetAllocator
+
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    hash::{DefaultHasher, Hash, Hasher},
+    iter,
+    sync::{Arc, RwLock},
+};
+
+use bevy_app::{App, Plugin};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    schedule::IntoSystemConfigs,
+    system::{ResMut, Resource},
+};
+use bevy_time::common_conditions::on_timer;
+use bevy_utils::{hashbrown::HashMap, prelude::default, tracing::error, Duration};
+use offset_allocator::{Allocation, Allocator};
+use slotmap::{new_key_type, SlotMap};
+use wgpu::{
+    util::BufferInitDescriptor, BufferDescriptor, BufferUsages, DownlevelFlags, IndexFormat,
+};
+
+use crate::{
+    mesh::MeshVertexBufferLayoutRef,
+    render_resource::Buffer,
+    renderer::{RenderAdapter, RenderDevice, RenderQueue},
+    Render, RenderApp,
+};
+
+/// How often we sweep unused allocations.
+const SWEEP_INTERVAL: Duration = Duration::from_secs(10);
+
+/// The default size of a slab, in bytes.
+const DEFAULT_SLAB_SIZE: u64 = 32 * 1024 * 1024;
+
+/// A plugin that provides the GPU memory allocator.
+pub struct GpuAllocatorPlugin {
+    /// The size of a slab.
+    ///
+    /// By default, this is 32 MB.
+    pub slab_size: u64,
+}
+
+/// Manages allocations in GPU buffers.
+#[derive(Resource, Clone)]
+pub struct GpuAllocator {
+    slabs: HashMap<SlabId, Buffer>,
+    slab_size: u64,
+    next_slab_id: SlabId,
+    classes: HashMap<GpuAllocationClass, GpuClassAllocator>,
+    adapter_downlevel_flags: DownlevelFlags,
+}
+
+#[derive(Clone, Default, Deref, DerefMut)]
+struct GpuClassAllocator(Arc<RwLock<GpuClassAllocatorData>>);
+
+#[derive(Default)]
+struct GpuClassAllocatorData {
+    regular_slabs: SlotMap<RegularSlabId, (Allocator, SlabId)>,
+    large_slabs: SlotMap<LargeSlabId, SlabId>,
+    free_large_slabs: Vec<SlabId>,
+}
+
+/// The type of a GPU buffer. Each class has its own allocator.
+///
+/// Only allocations of the same class can coexist in a single buffer.
+///
+/// Unlike regular CPU memory, GPU buffers require allocation with specific
+/// *usages*, which restrict how they can be used. Additionally, the APIs place
+/// additional restrictions: for example, because drawcalls require us to
+/// specify the initial vertex *index*, and not the initial vertex *byte
+/// position*, we must only group meshes with identical vertex buffer layouts
+/// into the same buffer.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum GpuAllocationClass {
```

Comment:
Curious if this makes sense to add other `GpuAllocationClass` textures, samplers, pipeline layout, etc?

Reply:
I think we should add those as we come to them. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) :)